bundle/mutator_read_only.go

```diff
@@ -2,28 +2,59 @@ package bundle
 
 import (
 	"context"
+	"sync"
 
 	"github.com/databricks/cli/libs/diag"
 	"github.com/databricks/cli/libs/log"
 )
 
-// ReadOnlyMutator is the interface type that allows access to bundle configuration but does not allow any mutations.
 type ReadOnlyMutator interface {
-	// Name returns the mutators name.
-	Name() string
+	Mutator
 
-	// Apply access the specified read-only bundle object.
-	Apply(context.Context, ReadOnlyBundle) diag.Diagnostics
+	// This is just tag, to differentiate this interface from bundle.Mutator
+	// This prevents non-readonly mutators being passed to ApplyParallel().
+	IsRO()
 }
 
-func ApplyReadOnly(ctx context.Context, rb ReadOnlyBundle, m ReadOnlyMutator) diag.Diagnostics {
-	ctx = log.NewContext(ctx, log.GetLogger(ctx).With("mutator (read-only)", m.Name()))
+// Helper to mark the mutator as "read-only"
+type RO struct{}
 
-	log.Debugf(ctx, "ApplyReadOnly")
-	diags := m.Apply(ctx, rb)
-	if err := diags.Error(); err != nil {
-		log.Debugf(ctx, "Error: %s", err)
+func (*RO) IsRO() {}
+
+// Run mutators in parallel. Unlike Apply and ApplySeq, this does not perform sync between
+// dynamic and static configuration.
+// Warning: none of the mutators involved must modify bundle directly or indirectly. In particular,
+// the must not call bundle.Apply or bundle.ApplySeq because those include writes to config even if mutator does not.
+// Deprecated: do not use for new use cases. Refactor your parallel task not to depend on bundle at all.
+func ApplyParallel(ctx context.Context, b *Bundle, mutators ...ReadOnlyMutator) diag.Diagnostics {
```

Comment:
I am concerned about this being a strictly worse interface. Earlier, we had at least some type system guarantees (because the bundle configuration was exposed via a function) that the bundle is not being modified, which goes away here.

I get that a deep-copy approach would offset this concern. Is this PR just needed for the migration itself?

Reply:
> Without deep-copy there are no guarantees.

There are no guarantee for pointer values. We do have guarantees for non-pointer values though (assuming none of the intermediate nodes are pointers if we are being strict).

Reply:
> not sure I understand.

I mean the migration to a deep copy approach.

Reply:
I think we should refactor the parallel operations so that deep-copying is not necessary. That's it, they should not be accepting full bundle.